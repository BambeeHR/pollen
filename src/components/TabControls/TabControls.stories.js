import CounterBadge from '../CounterBadge/CounterBadge.vue';
import TabControls from './TabControls.vue';

export default {
  title: 'TabControls',
  component: TabControls,
};

export const Default = () => ({
  components: { TabControls },
  data() {
    return {
      options: ['Foo', 'Bar', 'Bizz', 'Buzz'],
      selected: null,
    };
  },
  template: `<TabControls :options="options" v-model="selected" />`,
});

export const WithInitialValue = () => ({
  components: { TabControls },
  data() {
    return {
      options: ['Foo', 'Bar', 'Bizz', 'Buzz'],
      selected: 'Bar',
    };
  },
  template: `<TabControls :options="options" v-model="selected" />`,
});

export const FullWidthTabs = () => ({
  components: { TabControls },
  data() {
    return {
      options: ['Foo', 'Bar', 'Bizz', 'Buzz'],
      selected: null,
    };
  },
  template: `<TabControls :options="options" v-model="selected" full-width />`,
});

export const Wrapping = () => ({
  components: { TabControls },
  data() {
    return {
      options: ['Foo', 'Bar', 'Bizz', 'Buzz', 'Lorem Ipsum', 'Dolor Sit Amet'],
      selected: null,
    };
  },
  template: `<div class="max-w-xs"><TabControls :options="options" v-model="selected" full-width /></div>`,
});

export const WithLabelsAsComponents = () => ({
  components: { TabControls },
  data() {
    return {
      firstCount: 2,
      secondCount: 12,
      selected: null,
    };
  },
  computed: {
    options() {
      return [
        {
          label: {
            components: { CounterBadge },
            template: `<span class="flex items-center">Foo <CounterBadge :value="${this.firstCount}" class="ml-1" /></span>`,
          },
          value: 'foo',
        },
        {
          label: {
            components: { CounterBadge },
            template: `<span class="flex items-center">Bar <CounterBadge :value="${this.secondCount}" class="ml-1" /></span>`,
          },
          value: 'bar',
        },
      ];
    },
  },
  template: `<TabControls :options="options" v-model="selected" />`,
});
