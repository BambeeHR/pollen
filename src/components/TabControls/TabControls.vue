<template>
  <div class="tab-controls">
    <ul
      class="tab-controls__tabs"
      :class="{ 'tab-controls--full-width': fullWidth }"
      role="tablist"
    >
      <li
        v-for="item in mappedOptions"
        :key="item.value"
        :ref="item.value"
        class="tab-controls__tab"
        :class="{ 'tab-controls__tab--active': item.selected }"
      >
        <a
          :id="item.id"
          role="tab"
          :href="`#${item.target}`"
          :aria-selected="item.selected"
          :aria-controls="item.target"
          class="tab-controls__link"
          @click.prevent="$emit('input', item.value)"
        >
          <component :is="item.label" v-if="typeof item.label === 'object'" />
          <template v-else>{{ item.label }}</template>
        </a>
      </li>
    </ul>
    <div
      class="tab-controls__indicator"
      role="presentation"
      :style="indicatorStyle"
    />
  </div>
</template>

<script>
import kebabCase from 'lodash/kebabCase';
import shortid from 'shortid';
import mapOptions from '../../util/mapOptions';

export default {
  props: {
    fullWidth: {
      type: Boolean,
      default: false,
    },
    options: {
      type: Array,
      default: () => [],
    },
    tabsetId: {
      type: String,
      default: shortid.generate,
    },
    value: {
      type: String,
      default: null,
    },
  },
  data() {
    return {
      indicatorStyle: {},
    };
  },
  computed: {
    mappedOptions() {
      return mapOptions(this.options).map((item) => ({
        ...item,
        target: `${this.tabsetId}-${kebabCase(item.value).toLowerCase()}-panel`,
        id: `${this.tabsetId}-${kebabCase(item.value).toLowerCase()}-tab`,
        selected: item.value === this.value,
      }));
    },
  },
  watch: {
    value() {
      this.positionIndicator();
    },
  },
  mounted() {
    this.positionIndicator();
  },
  methods: {
    positionIndicator() {
      const activeLabel = this.$refs[this.value]
        ? this.$refs[this.value][0]
        : null;
      let indicatorStyle = {};
      if (activeLabel) {
        indicatorStyle = {
          left: `${activeLabel.offsetLeft}px`,
          top: `${activeLabel.offsetTop + activeLabel.offsetHeight}px`,
          width: `${activeLabel.offsetWidth}px`,
        };
      }
      this.indicatorStyle = indicatorStyle;
    },
  },
};
</script>

<style scoped>
.tab-controls {
  @apply relative;
}

.tab-controls__tabs {
  @apply border-b 
    border-l-0 
    border-r-0 
    border-t-0 
    border-gray-5 
    border-solid 
    flex 
    flex-wrap 
    list-none 
    m-0
    p-0;
}

.tab-controls--full-width .tab-controls__tab {
  @apply flex-grow text-center;
}

.tab-controls__link {
  @apply block
    border-b 
    border-l-0 
    border-r-0 
    border-t-0
    border-transparent
    border-solid 
    font-body-medium 
    font-medium 
    no-underline 
    px-4 
    py-2 
    text-gray-0;
}

.tab-controls__indicator {
  @apply absolute 
  bg-primary 
    duration-light
    ease-out
    h-px;
  transition-property: left, width;
}
</style>
