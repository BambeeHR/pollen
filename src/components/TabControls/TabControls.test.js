import { mount } from '@vue/test-utils';
import TabControls from './TabControls.vue';

const options = [
  'Option 1',
  {
    label: 'Option 2',
    value: 'foo',
  },
  {
    label: 'Option 3',
    value: 'bar',
  },
];

describe('TabControls', () => {
  test('renders correctly', () => {
    const wrapper = mount(TabControls, {
      propsData: {
        options,
        tabsetId: 'test',
      },
    });
    expect(wrapper.element).toMatchSnapshot();
  });

  test('highlights the correct label', async () => {
    const wrapper = mount(TabControls);
    wrapper.setProps({
      options,
      value: 'foo',
    });
    await wrapper.vm.$nextTick();
    expect(wrapper.find('.tab-controls__tab--active').text()).toBe('Option 2');
  });

  test('updates bgStyle correctly', () => {
    const expectedStyle = {
      left: '10px',
      top: '20px',
      width: '10px',
    };
    const wrapper = mount(TabControls, {
      propsData: {
        options,
        value: 'foo',
      },
    });
    wrapper.vm.$refs = {
      foo: [
        {
          offsetLeft: 10,
          offsetHeight: 10,
          offsetTop: 10,
          offsetWidth: 10,
        },
      ],
    };
    wrapper.vm.positionIndicator();
    expect(wrapper.vm.indicatorStyle).toEqual(expectedStyle);
  });
});
