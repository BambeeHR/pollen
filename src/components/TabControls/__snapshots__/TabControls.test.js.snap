// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TabControls renders correctly 1`] = `
<div
  class="tab-controls"
>
  <ul
    class="tab-controls__tabs"
    role="tablist"
  >
    <li
      class="tab-controls__tab"
    >
      <a
        aria-controls="test-option-1-panel"
        class="tab-controls__link"
        href="#test-option-1-panel"
        id="test-option-1-tab"
        role="tab"
      >
        Option 1
      </a>
    </li>
    <li
      class="tab-controls__tab"
    >
      <a
        aria-controls="test-foo-panel"
        class="tab-controls__link"
        href="#test-foo-panel"
        id="test-foo-tab"
        role="tab"
      >
        Option 2
      </a>
    </li>
    <li
      class="tab-controls__tab"
    >
      <a
        aria-controls="test-bar-panel"
        class="tab-controls__link"
        href="#test-bar-panel"
        id="test-bar-tab"
        role="tab"
      >
        Option 3
      </a>
    </li>
  </ul>
   
  <div
    class="tab-controls__indicator"
    role="presentation"
  />
</div>
`;
